### PR TITLE
update Sphinx documentation

### LINKED ISSUE #24: Update Sphinx documentation
While a rudimentary sphinx documentation is now implemented, it is not sufficiently detailed.

To-do:
- [x] Move `README.md` to `REAMDE.rst` and update syntax
- [x] Create separate `INSTALL.rst` file
- [x] Tie `README.rst` and `INSTALL.rst` into sphinx docs
- [x] Add usage examples to module/class docstrings
- [x] Update jupyter notebooks and add more of them to sphinx
- [x] Add additional discussion on how things work
- [x] Switch to alabaster theme?

`camsym`'s documentation is probably one of the better examples of where this should go.

### PR DESCRIPTION
**Pull request by DavidAkinpelu**
Closes #24

### CODE REVIEW

**Comment by ischoegl** [comment]
While I'll provide more feedback tomorrow, I believe you didn't delete the old jupyter examples, right? They probably should go, as they're replicated in docs/pages?

File: ctwrap/simulation.py [hunk 1]
```diff
@@ -1,4 +1,11 @@
-"""Object handling variations."""
+"""This ctwraper module handles batch cantera simulations variations.
+Cantera based modules are pass to this module where they are run in batches.
+see `adiabatic flame example <../examples/adiabatic_flame_example.ipynb>`_ or
+`ignition_example <../examples/ignition_example.ipynb>`_ .
```
Comment (ischoegl) [inline ctwrap/simulation.py:4]: Make sure those links work once sphinx has processed this. Also, they should point to the notebooks in the pages folder  (the notebooks in examples are redundant and should be removed).

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+The code was developed with the following objectives in mind. It should:
+
+ * create a flexible framework for generic simulations
+ * provide a command line interface
+ * be easily scriptable (using YAML configuration files via `ruamel.yaml`)
+ * enforce units (via `pint`)
+ * avoid clutter (data are saved in HDF containers)
+ * results can be easily processed (via `pandas` HDF implementation)
+ * generate self-documenting results (save configuration with results)
+ * enable parallel execution of (single-threaded) simulations (via `multiprocessing`)
```
Comment (ischoegl) [inline README.rst:60]: Please revise this list - see Cantera/enhancements#69

File: ctwrap/simulation.py [hunk 3]
```diff
@@ -27,7 +34,10 @@
 class Simulation(object):
     """
     The Simulation class wraps modules into an object.
-    Required module functions 'run' and 'defaults' are passed through
+    Required module functions 'run', 'save' and 'defaults' are passed through
```
Comment (ischoegl) [inline ctwrap/simulation.py:37]: use double backticks instead of `'` so these appear monospaced.

File: README.rst [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ctwrap
+======
+
+Light-weight Python wrapper for batch simulation jobs (e.g. Cantera). The source
+code is available at the `ctwrap Github repositiory <https://github.com/microcombustion/ctwrap/>`_.
+
+++++++++
+Overview
+++++++++
+Ctwrap is a light-weight batch simulation wrapper written in Python. The wrapper consists of two modules:
```
Comment (ischoegl) [inline README.rst:16]: Leave an empty line between title and paragraph

File: ctwrap/simulation.py [hunk 5]
```diff
@@ -59,15 +69,21 @@ def __init__(self, module: str, output: Dict[str, Any] = None):
...
-        The `from_module` instantiation call renames the class
-        based on the module name.
+
+        The :meth:`from_module` method is intended as the main route for the creation of
+        `Simulation` objects. For example:
+
+        .. code-block:: Python
+
+           # creates a simulation object with module `minimal`
+           sim = ctwrap.Simulation.from_module(cw.modules.minimal)
```
Comment (ischoegl) [inline ctwrap/simulation.py:82]: replace `cw` by `ctwrap` ... `cw` is only clear if you `import ctwrap as cw`

File: ctwrap/simulation.py [hunk 6]
```diff
@@ -59,15 +69,21 @@ def __init__(self, module: str, output: Dict[str, Any] = None):
...
 
         Arguments:
            module (module): handle name or handle to module running the simulation
-           output (dict): generated by `SimulationHandler`
+           output (dict):  Output information. Generated by `SimulationHandler`
```
Comment (ischoegl) [inline ctwrap/simulation.py:86]: The `output` probably needs some additional discussion (in #28?) as only HDF is left.

File: ctwrap/simulation.py [hunk 7]
```diff
@@ -84,7 +100,15 @@ def run(self, name: Optional[str] ='defaults',
             config: Optional[Dict[str, Any]] = None,
             **kwargs: str) -> None:
         """
-        Run function holding configuration in dictionary.
+        Runs the simulation module's run method.
+
+        The :meth:`run` is used to call the module's run method.
+        A simple usage is
```
Comment (ischoegl) [inline ctwrap/simulation.py:106]: be more specific. Also, there are some nouns missing ... explicitly write 'method' after 
```
:meth:`run`
```
and add 'example' after usage.

File: ctwrap/simulation.py [hunk 8]
```diff
@@ -84,7 +100,15 @@ def run(self, name: Optional[str] ='defaults',
...
+
+        .. code-block:: Python
+
+           # calls the run method in module 'minimal`
```
Comment (ischoegl) [inline ctwrap/simulation.py:110]: I think every time you mention module, replace it by 'simulation module' for clarity

File: ctwrap/simulation.py [hunk 9]
```diff
@@ -149,7 +175,11 @@ def _save_metadata(self,
                        metadata: Optional[Dict[str, Any]] = None,
                        ) -> None:
         """
-        This function calls the module save method.
+        Modules saves the input file as an attribute after the
+        batch simulation has been completed
+
+        This method calls the
+        :py:func:`~ctwrap.parser.save_metadata`.
```
Comment (ischoegl) [inline ctwrap/simulation.py:182]: add 'method' at the end of the sentence.

File: ctwrap/simulation.py [hunk 10]
```diff
@@ -168,6 +198,25 @@ class SimulationHandler(object):
     """
     Class handling parameter variations.
     Class adds methods to switch between multiple configurations.
+
+    .. note:: :class:`SimulationHandler` objects should be
+        instantiated using the :meth:`from_yaml`.
```
Comment (ischoegl) [inline ctwrap/simulation.py:203]: add 'method' at the end of the sentence.

File: ctwrap/simulation.py [hunk 11]
```diff
@@ -168,6 +198,25 @@ class SimulationHandler(object):
...
+
+    An example of a minimal yaml input is:
```
Comment (ischoegl) [inline ctwrap/simulation.py:205]: replace 'input' by 'configuration'

File: ctwrap/simulation.py [hunk 12]
```diff
@@ -168,6 +198,25 @@ class SimulationHandler(object):
...
+
+       .. code-block:: YAML
+
+          output: # output file information
+            format: null # no file output
```
Comment (ischoegl) [inline ctwrap/simulation.py:210]: We should probably drop this as it is superseded by the `save` methods that are now moved to the simulation modules.

File: ctwrap/simulation.py [hunk 13]
```diff
@@ -393,18 +448,28 @@ def output_name(self):
...
         return {'{}_{}'.format(e, v): v for v in self._values}
 
     def run_task(self, sim: Simulation, task: str, **kwargs:str):
         """
         Function to run a specific task.
 
+        The :meth:`run_task` method calls the module's run and save method to run and
+         save the resulting output object and also afterwards calls
+         :py:func:`~ctwrap.parser._save_metadata`. to save the metadata.
+         A simple usage is:
```
Comment (ischoegl) [inline ctwrap/simulation.py:466]: missing `example`

File: ctwrap/simulation.py [hunk 14]
```diff
@@ -393,18 +448,28 @@ def output_name(self):
...
         e = self._entry
         return {'{}_{}'.format(e, v): v for v in self._values}
 
     def run_task(self, sim: Simulation, task: str, **kwargs:str):
         """
         Function to run a specific task.
 
+        The :meth:`run_task` method calls the module's run and save method to run and
+         save the resulting output object and also afterwards calls
+         :py:func:`~ctwrap.parser._save_metadata`. to save the metadata.
```
Comment (ischoegl) [inline ctwrap/simulation.py:465]: 'Data' is plural, so you can drop the article 'the' in front of metadata.

File: ctwrap/simulation.py [hunk 15]
```diff
@@ -431,11 +496,18 @@ def run_serial(self,
         """
         Run variation in series.
 
+        The :meth:`run_serial` method runs all the variations in the input
+        file serially. A simple usage is:
```
Comment (ischoegl) [inline ctwrap/simulation.py:500]: example

File: ctwrap/simulation.py [hunk 16]
```diff
@@ -471,12 +543,19 @@ def run_parallel(self,
         """
         Run variation using multiprocessing.
 
+        The :meth:`run_parallel` method runs all the variations in the input
+        file in parallel using python multiprocessing. A minimal example is:
```
Comment (ischoegl) [inline ctwrap/simulation.py:547]: used single back-ticks around multiprocessing and see whether this creates a link. Adding `:any:` in front of it may be necessary.

File: ctwrap/simulation.py [hunk 17]
```diff
@@ -471,12 +543,19 @@ def run_parallel(self,
...
+
+         .. code-block:: Python
+
+           # Runs all the variations in parallel
+           sh.run_parallel(sim)
```
Comment (ischoegl) [inline ctwrap/simulation.py:552]: whenever you use a variable name (like `sh` here), make sure that it is mentioned in the text describing the example.

File: ctwrap/simulation.py [hunk 18]
```diff
@@ -533,7 +612,7 @@ def worker(tasks_to_accomplish, tasks_that_are_done, module: str, lock,
            metadata: Dict[str, Any],
            verbosity: int) -> True:
     """
-    Worker function.
+    Worker function for the `:meth: run_parallel` method.
```
Comment (ischoegl) [inline ctwrap/simulation.py:615]: misplaced back ticks?

File: docs/index.rst [hunk 19]
```diff
@@ -6,6 +6,12 @@
 ctwrap - Python Wrapper for Batch Simulations
 =============================================
 
+.. include:: ../README.rst
+
+.. include:: ../install.rst
+
+.. include:: ../CONTRIBUTING.rst
```
Comment (ischoegl) [inline docs/index.rst:13]: I'd prefer to use dedicated pages for these, e.g. see `camsym` for installation. That way the page remains at a manageable length. The `install.rst` should probably be `INSTALL.rst` in the root folder. For whatever reason, it appears to be custom to do so.

File: docs/index.rst [hunk 20]
```diff
@@ -6,6 +6,12 @@
...
+
 .. toctree::
    :maxdepth: 2
    :caption: Contents:
```
Comment (ischoegl) [inline docs/index.rst:17]: Please leave the table of contents at the top of the page (it is very useful for navigation).

File: docs/pages/jupyter.rst [hunk 21]
```diff
@@ -5,3 +5,7 @@ Jupyter Notebooks
     :maxdepth: 2
 
     Parser Example <parser_example.ipynb>
+	Adiabatic Flame Example <adiabatic_flame_example.ipynb>
+	Ignition Example <ignition_example.ipynb>
+	Commandline Example <commandline_example.ipynb>
+	Minimal Example <minimal_example.ipynb>
```
Comment (ischoegl) [inline docs/pages/jupyter.rst:11]: check indent level.

File: install.rst [hunk 22]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (ischoegl) [inline install.rst:1]: No empty lines at top.

File: install.rst [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+
+
++++++++++
+Uninstall
++++++++++
+::
+
+    $ pip uninstall ctwrap
+
```
Comment (ischoegl) [inline install.rst:37]: Files should end with a single empty line.

**Review by ischoegl — CHANGES_REQUESTED** [review]
This looks quite good, although I have a couple of requsts.

I haven't looked at the jupyter notebooks yet. Please remove the old `examples` that are now redundant. I may have more comments on those which I'll leave separately.

**Comment by ischoegl** [comment]
Ad sphinx documentation - note that you can download what the GitHub action generates as an 'artifact' (it's attached to the check - under the 'Checks' tab at the top of the page).

While the example that I added shows output, yours do not (i.e. no plots etc.) - please execute the notebooks, and save them with graphs embedded. For the minimal example, you can remove the yaml input and the simulation module ... I believe discussing those in the `Batch Job` section is easier, as much of that content is auto-generated and thus doesn't need to be updated every time something changes.

File: ctwrap/simulation.py (hunk 7 above)
Comment (DavidAkinpelu) [reply ctwrap/simulation.py:106]: An example is on the next line

File: ctwrap/simulation.py (hunk 12 above)
Comment (DavidAkinpelu) [reply ctwrap/simulation.py:210]: I understand while we need to drop `format` but  `output` should not be entirely dropped so the user can specify `filename`

File: ctwrap/simulation.py (hunk 13 above)
Comment (DavidAkinpelu) [reply ctwrap/simulation.py:466]: The code block is below

File: ctwrap/simulation.py (hunk 7 above)
Comment (ischoegl) [reply ctwrap/simulation.py:106]: there is ... what I meant is that you should write 'A simple usage example is'

File: ctwrap/simulation.py (hunk 12 above)
Comment (ischoegl) [reply ctwrap/simulation.py:210]: ok.

File: ctwrap/simulation.py (hunk 7 above)
Comment (DavidAkinpelu) [reply ctwrap/simulation.py:106]: Oh!

**Comment by ischoegl** [comment]
Please leave `INSTALL.rst` and `CONTRIBUTING.rst` in the root folder. Also, to make things simpler for review, please mark things you addressed as 'resolved' (if there's any context needed, please respond. If you adopt my recommendations, you can just respond with a :+1: reaction.

File: docs/index.rst (hunk 19 above)
Comment (ischoegl) [reply docs/index.rst]: Please address / also see other comment.

File: README.rst (hunk 2 above)
Comment (ischoegl) [reply README.rst]: My comment was mostly about the list, not the rest of the file (some of the changes in 0.2.0 made the earlier version outdated).

File: docs/overview.rst [hunk 24]
```diff
@@ -0,0 +1,29 @@
+========
+Overview
+========
+
+Ctwrap is a light-weight batch simulation wrapper written in Python. The wrapper consists of two modules:
+the ``parser`` and the ``simulation`` modules. The ``parser`` module is used to parsed the input file and also
```
Comment (ischoegl) [inline docs/overview.rst:6]: Could you rephrase this in terms of objects/classes rather than modules? I believe the important things here are `Parser`, `Simulation` and `SimulationHandler`. You mention this further down, but it may be better to skip to the important part immediately.

File: docs/overview.rst [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+handles units using `pint <https://pint.readthedocs.io/en/stable/>`_,  while
+the ``simulation`` modules handles running batch operation. The ``simulation`` module consists of two classes:
+the :py:class:`~ctwrap.simulation.Simulation` that wraps a simulation module to be run.
```
Comment (ischoegl) [inline docs/overview.rst:9]: :py:class:`~ctwrap.simulation.Simulation` can be replaced by :any:`Simulation`.

File: docs/overview.rst [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+More details about the module is discussed below and the
+:py:class:`~ctwrap.simulation.SimulationHandler` that handles parameter variations and switches between
+multiple configurations.
+
+The wrapper requires a simulation module, a single run python module, that should consists of three
```
Comment (ischoegl) [inline docs/overview.rst:14]: Comment that this isn't much different from a conventional script. All that is done is that you 'organize' the code

File: docs/overview.rst [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+methods: ``defaults``, ``run`` and ``save``. See `minimal module <pages/minimal.py>`_.
+The ``defaults`` method should return the default input parameters.
+The ``run`` method runs a single simulation and returns a dictionary containing the output object.
+The save method saves the output object based on the specified
+format in the yaml input file.
+
+The inputs are passed through a yaml file. The yaml is structured into three parts: ``output``,
```
Comment (ischoegl) [inline docs/overview.rst:21]: I think uppercase YAML is easier to capture visually

File: docs/overview.rst [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+``variation`` and ``defaults``. The ``output`` contains information about the resulting file to be
+saved  such as the ``file_name``, ``path`` etc. The ``file_name`` is optional, and in that case,
```
Comment (ischoegl) [inline docs/overview.rst:23]: Some of this is much easier explained if you have an example on hand. It may make sense to merge the `minimal.rst` page with this overview?

File: docs/overview.rst [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+the file is saved with the name of the module to be run. The ``variation`` contains ``entry``
+and ``values`` to be varied, while the ``defaults`` contains the defaults input required to run
+the module see `minimal yaml <pages/minimal.yaml>`_.
+
+
+*Note*: ``ctwraper`` only saves in `h5` format.
```
Comment (ischoegl) [inline docs/overview.rst:29]: ``ctwrap`` ?

File: README.rst [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+
+    $ ctwrap adiabatic_flame adiabatic_flame.yaml --parallel
+
+Results are written to a single file ``adiabatic_flame.h5``.
+
+Adapting custom simulation tasks to be used with ctwrap is straight-forward,
+and mainly involves the creation of a Python module that largely reflects a
+regular simulation script for a single set of parameter values.
+
+*Alternatives*
```
Comment (ischoegl) [inline README.rst:49]: Please delete the next paragraphs - they were meant as comments in the discussion on Cantera/enhancements

**Comment by ischoegl** [comment]
While there are a couple of small issues remaining, I’ll merge. It’s easier to discuss once the updated Sphinx docs are online.
